lib/Sema/DebuggerTestingTransform.cpp

```diff
@@ -0,0 +1,245 @@
+    if (!DstDRE)
+      return {true, OriginalExpr};
+
+    ValueDecl *DstDecl = DstDRE->getDecl();
+    if (!DstDecl->hasName())
+      return {true, OriginalExpr};
+
+    // Don't capture variables which aren't default-initialized.
+    if (auto *VD = dyn_cast<VarDecl>(DstDecl))
+      if (!VD->getParentInitializer() && !VD->isInOut())
+        return {true, OriginalExpr};
+
+    // Rewrite the original expression into this:
+    // call
+    //   closure {
+    //     $OriginalExpr
+    //     checkExpect("$Varname", _stringForPrintObject($Varname))
+    //   }
+
+    // Create "$Varname".
+    llvm::SmallString<256> DstNameBuf;
+    DeclName DstDN = DstDecl->getFullName();
+    StringRef DstName = Ctx.AllocateCopy(DstDN.getString(DstNameBuf));
+    assert(!DstName.empty() && "Varname must be non-empty");
+    Expr *Varname = new (Ctx) StringLiteralExpr(DstName, SourceRange());
+    Varname->setImplicit(true);
+
+    // Create _stringForPrintObject($Varname).
+    auto *PODeclRef = new (Ctx)
+        UnresolvedDeclRefExpr(Ctx.getIdentifier("_stringForPrintObject"),
+                              DeclRefKind::Ordinary, DeclNameLoc());
+    Expr *POArgs[] = {DstDRE};
+    Identifier POLabels[] = {Identifier()};
+    auto *POCall = CallExpr::createImplicit(Ctx, PODeclRef, POArgs, POLabels);
+    POCall->setThrows(false);
+
+    // Create the call to checkExpect.
+    Identifier CheckExpectLabels[] = {Identifier(), Identifier()};
+    Expr *CheckExpectArgs[] = {Varname, POCall};
+    UnresolvedDeclRefExpr *CheckExpectDRE = new (Ctx)
+        UnresolvedDeclRefExpr(Ctx.getIdentifier("_debuggerTestingCheckExpect"),
+                              DeclRefKind::Ordinary, DeclNameLoc());
+    auto *CheckExpectExpr = CallExpr::createImplicit(
+        Ctx, CheckExpectDRE, CheckExpectArgs, CheckExpectLabels);
+    CheckExpectExpr->setThrows(false);
+
+    // Create the closure.
+    TypeChecker TC{Ctx};
+    auto *Params = ParameterList::createEmpty(Ctx);
+    auto *Closure = new (Ctx)
+        ClosureExpr(Params, SourceLoc(), SourceLoc(), SourceLoc(), TypeLoc(),
+                    DF.getNextDiscriminator(), getCurrentDeclContext());
+    Closure->setImplicit(true);
+
+    // TODO: Save and return the value of $OriginalExpr.
+    ASTNode ClosureElements[] = {OriginalExpr, CheckExpectExpr};
+    auto *ClosureBody = BraceStmt::create(Ctx, SourceLoc(), ClosureElements,
+                                          SourceLoc(), /*Implicit=*/true);
+    Closure->setBody(ClosureBody, /*isSingleExpression=*/false);
+
+    // Call the closure.
+    auto *ClosureCall = CallExpr::createImplicit(Ctx, Closure, {}, {});
+    ClosureCall->setThrows(false);
+
+    // TODO: typeCheckExpression() seems to assign types to everything here,
+    // but may not be sufficient in some cases.
+    Expr *FinalExpr = ClosureCall;
+    if (!TC.typeCheckExpression(FinalExpr, getCurrentDeclContext()))
+      llvm::report_fatal_error("Could not type-check instrumentation");
```

Comment:
Ditto, if we create the instrumentation ourselves, shouldn't we be able to guarantee always that the type checking succeeds? (and if not, that is a bug, I guess)

Reply:
We should demand that the type checking always succeed, hence the fatal error. I'm fine with using an assert, but that seems less strict.